src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -467,58 +484,113 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
 
         private static Func<object?, HttpContext, Task> HandleRequestBodyAndCompileRequestDelegate(Expression responseWritingMethodCall, FactoryContext factoryContext)
         {
-            if (factoryContext.JsonRequestBodyType is null)
+            if (factoryContext.JsonRequestBodyType is null && !factoryContext.ReadForm)
             {
                 return Expression.Lambda<Func<object?, HttpContext, Task>>(
                     responseWritingMethodCall, TargetExpr, HttpContextExpr).Compile();
             }
 
-            // We need to generate the code for reading from the body before calling into the delegate
+            // We need to generate the code for reading from the body or form before calling into the delegate
             var invoker = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(
                 responseWritingMethodCall, TargetExpr, HttpContextExpr, BodyValueExpr).Compile();
 
             var bodyType = factoryContext.JsonRequestBodyType;
-            object? defaultBodyValue = null;
+            object? defaultContentValue = null;
 
-            if (factoryContext.AllowEmptyRequestBody && bodyType.IsValueType)
+            if (!factoryContext.ReadForm && factoryContext.AllowEmptyRequestBody && bodyType is { IsValueType: true })
             {
-                defaultBodyValue = Activator.CreateInstance(bodyType);
+                defaultContentValue = Activator.CreateInstance(bodyType);
             }
 
             return async (target, httpContext) =>
             {
-                object? bodyValue = defaultBodyValue;
+                object? contentValue = defaultContentValue;
+
+                var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+                if (feature?.CanHaveBody == true)
+                {
+                    if (bodyType is not null)
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFromJsonAsync(bodyType);
+                        }
+                        catch (IOException ex)
+                        {
+                            Log.RequestBodyIOException(httpContext, ex);
+                            return;
+                        }
+                        catch (InvalidDataException ex)
+                        {
+                            Log.RequestBodyInvalidDataException(httpContext, ex);
+                            httpContext.Response.StatusCode = 400;
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFormAsync();
```

Comment:
This will throw `InvalidOperationException` if the content-type is incorrect. Should that be caught and an HTTP 415 status returned?

https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Http/Http/src/Features/FormFeature.cs#L148-L151

Reply:
9d71e8a483fe43c3620bac119b7906b670ee8c9a